front/lib/api/assistant/agent.ts

```diff
@@ -596,6 +599,16 @@ async function* runAction(
 > {
   const now = Date.now();
 
+  const existingRetrievalActions = agentMessage.actions.filter(
```

Comment:
Wait can't this happen in the near future concurrently on two retrieval actions?

Reply:
Yes for concurrency it will be a thing (but we don't have it yet)

Reply:
After meditating, I think a more principled approach would be to assign a rank to each action.

This is reusable in other context and we can allocate a band of references per rank without running out of them (with maybe an error if we go beyond a number for the retrieval action rank)

We may not need to save that rank yet, just need to assign it at execution of the action

This will extend nicely to parallel action.

WDYT?

Reply:
Agreed re testing phase but it's also in these initial building phase that we lay the landscape for the months if not years to come. Hence why I'm looking for inverted 80/20 approaches in such central places

Reply:
Very fair, will fix

Reply:
Awesome!